File: leetcode/solutions/lru-cache.swift
class Node {
    var value: Int?
    var key: Int?
    var next: Node?
    var prev: Node?
}

class LRUCache {
    var cache = [Int: Node]()
    var size = 0
    let capacity: Int
    let head = Node()
    let tail = Node()

    init(_ capacity: Int) {
        self.capacity = capacity
        head.next = tail
        tail.prev = head
    }

    func get(_ key: Int) -> Int {
        guard
            let node = cache[key],
            let value = node.value else {
            return -1
        }
        moveToHead(node)
        return value
    }

    func put(_ key: Int, _ value: Int) {
        guard let node = cache[key] else {
            let new = Node()
            new.key = key
            new.value = value
            cache[key] = new
            add(new)
            size += 1
            if size > capacity, let node = pop(), let key = node.key {
                cache.removeValue(forKey: key)
                size -= 1
            }
            return
        }
        node.value = value
        moveToHead(node)
    }

    private func add(_ node: Node) {
        node.prev = head
        node.next = head.next
        head.next?.prev = node
        head.next = node
    }

    private func remove(_ node: Node) {
        guard let prev = node.prev, let next = node.next else {
            return
        }
        prev.next = next
        next.prev = prev
    }

    private func moveToHead(_ node: Node) {
        remove(node)
        add(node)
    }

    private func pop() -> Node? {
        guard let node = tail.prev, node !== head else {
            return nil
        }
        remove(node)
        return node
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache(capacity)
 * let ret_1: Int = obj.get(key)
 * obj.put(key, value)
 */

File: leetcode/solutions/lru-cache.py
class Node():
    """
    Node
    """
    def __init__(self):
        self.key = None
        self.value = None
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.cache = {}
        self.size = 0
        self.capacity = capacity
        self.head, self.tail = Node(), Node()
        self.head.next = self.tail
        self.tail.prev = self.head


    def get(self, key: int) -> int:
        """
        get
        """
        node = self.cache.get(key, None)
        if not node:
            return None
        self._move_to_head(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        """
        put
        """
        node = self.cache.get(key)

        if not node:
            new = Node()
            new.key = key
            new.value = value
            self.cache[key] = new
            self._add(new)
            self.size += 1
            if self.size > self.capacity:
                tail = self._pop()
                del self.cache[tail.key]
                self.size -= 1
        else:
            node.value = value
            self._move_to_head(node)

    def _add(self, node):
        node.prev = self.head
        node.next = self.head.next
        self.head.next.prev = node
        self.head.next = node

    def _remove(self, node):
        prev = node.prev
        new = node.next
        prev.next = new
        new.prev = prev

    def _move_to_head(self, node):
        self._remove(node)
        self._add(node)

    def _pop(self):
        node = self.tail.prev
        self._remove(node)
        return node


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
